File: linda-core/src/main/java/it/unibo/coordination/linda/core/TupleSpace.java
package it.unibo.coordination.linda.core;

import it.unibo.coordination.utils.FutureUtils;
import org.apache.commons.collections4.MultiSet;

import java.util.concurrent.CompletableFuture;
import java.util.concurrent.Executor;

public interface TupleSpace<T extends Tuple, TT extends Template, K, V> {
    CompletableFuture<Match<T, TT, K, V>> read(TT template);

    default CompletableFuture<T> readTuple(TT template) {
        return FutureUtils.applyAndPropagateCancelAsync(
                read(template),
                match -> match.getTuple().get(),
                getExecutor()
        );
    }

    CompletableFuture<Match<T, TT, K, V>> take(TT template);

    default CompletableFuture<T> takeTuple(TT template) {
        return FutureUtils.applyAndPropagateCancelAsync(
                take(template),
                match -> match.getTuple().get(),
                getExecutor()
        );
    }

    CompletableFuture<T> write(T tuple);

    CompletableFuture<MultiSet<T>> get();

    CompletableFuture<Integer> getSize();

    String getName();

    Executor getExecutor();
}

File: linda-core/src/main/kotlin/it/unibo/coordination/linda/core/TupleSpace.kt
package it.unibo.coordination.linda.core

import it.unibo.coordination.linda.core.traits.*

interface TupleSpace<T : Tuple<T>, TT : Template<T>, K, V, M : Match<T, TT, K, V>>
    : Naming,
        OperationalPrimitives,
        LindaPrimitives<T, TT, K, V, M>,
        BulkPrimitives<T, TT, K, V, M>,
        NegatedPrimitives<T, TT, K, V, M>,
        PredicativePrimitives<T, TT, K, V, M>,
        NegatedPredicativePrimitives<T, TT, K, V, M>
